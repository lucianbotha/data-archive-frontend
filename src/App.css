.dimScreen {
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  height: 100%;
  width: 100%;
  margin: 0;
  padding: 0;
  background: #000000;
  opacity: 0.5;
  filter: alpha(opacity=30);
  -moz-opacity: 0.3;
  z-index: 101;
}
.signal {
  border: 15px solid #fff;
  border-radius: 30px;
  height: 130px;
  left: 50%;
  margin: -15px 0 0 -15px;
  opacity: 0;
  position: absolute;
  top: 50%;
  width: 130px;

  animation: pulsate 1s ease-out;
  animation-iteration-count: infinite;
}

@keyframes pulsate {
  0% {
    transform: scale(0.1);
    opacity: 0;
  }
  50% {
    opacity: 2;
  }
  100% {
    transform: scale(3);
    opacity: 0;
  }
}

.styles_modal__gNwvD {
  max-width: 1224px;
  width: 100%;
}

.cart-table {
  max-width: 1224px;
  overflow: auto;
  padding: 15px;
  height: 600px;
}
.c-table {
  border: 2px solid hsl(0, 0%, 86%);
}
.button-scales {
  width: 400px;
}
