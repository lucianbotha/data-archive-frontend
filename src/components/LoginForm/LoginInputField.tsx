import * as React from "react";
import InputField from "../basicComponents/InputField";

/*
The login input field responsible for user input.

Properties:
-----------
label
  The label which detail the user input field
name
  The name that uniquecly identifies the input field
onChange
  The fuction that is responsible for updating the input field value when changed
placeholder
  It gives an example of the required value
type
  The type of the input field e.g. text, password, etc.
value
  The value of the input field
*/

interface ILoginInputProps {
  error?: string;
  label?: string;
  loading?: boolean;
  name?: string;
  onChange?: any;
  placeholder?: string;
  type?: string;
  value?: string;
}

const LoginInputField = (props: ILoginInputProps) => {
  const { error, label, loading, name, onChange, type, value } = props;
  return (
    <div className={"field"}>
      <label htmlFor={`${name}+${label}`} className={"label"}>
        {label}:
      </label>
      <div className={"control is-child"}>
        <InputField
          data-test="target-name-input"
          disabled={loading}
          name={name}
          value={value || ""}
          error={error}
          onChange={onChange}
          type={type}
        />
      </div>
    </div>
  );
};

export default LoginInputField;
