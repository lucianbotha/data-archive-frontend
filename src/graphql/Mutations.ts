import gql from "graphql-tag";

/**
 * Creating a new user.
 *
 * The following arguments must be supplied.
 *
 * affiliation:
 *     The affiliation of the user, such as a university or an institute.
 * email:
 *     Email address. This will be stored in lower case.
 * familyName:
 *     The family name (surname).
 * givenName:
 *     The given name (first name).
 * password:
 *     The password, which must have at least 7 characters.
 * username:
 *     The username, which must not contain upper case letters.
 */
const SIGNUP_MUTATION = gql`
  mutation SIGNUP_MUTATION(
    $familyName: String!
    $givenName: String!
    $username: String!
    $email: String!
    $affiliation: String!
    $password: String!
  ) {
    signup(
      familyName: $familyName
      givenName: $givenName
      username: $username
      email: $email
      affiliation: $affiliation
      password: $password
    ) {
      id
    }
  }
`;

/**
 * Get currently logged in user.
 *
 * The following arguments must be supplied.
 * password:
 *     The correct user password.
 * username:
 *     The correct user username.
 */
const GET_USER_MUTATION = gql`
  mutation GET_USER_MUTATION($username: String!, $password: String!) {
    login(password: $password, username: $username) @client
  }
`;

export { GET_USER_MUTATION, SIGNUP_MUTATION };
