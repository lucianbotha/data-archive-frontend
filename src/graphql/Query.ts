import gql from "graphql-tag";

// TODO UPDATE
/**
 * Retrieving the user information.
 */
const USER_QUERY = gql`
  query USER_QUERY {
    user {
      givenName
    }
  }
`;

export { USER_QUERY };
