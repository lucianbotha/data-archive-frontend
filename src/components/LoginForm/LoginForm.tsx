import PropTypes from "prop-types";
import * as React from "react";
import styled from "styled-components";
import api from "../../api/api";
import { validateLoginField } from "../../util/LoginFormValidation";
import LoginInputField from "./LoginInputField";

/*
The login form which is responsible for authenticating th user into the data acrchive system.

Properties:
-----------
userInput : {
  password
    The password, which must have at least 7 characters.
  username
    The username, which must not contain upper case letters.
}

errors : []
  An array of error messages if any.
*/

interface ILoginFormState {
  errors: {
    password: string;
    responseError?: string;
    username: string;
  };
  userInput: {
    password: string;
    username: string;
  };
}

const LoginFormParent = styled.form.attrs({
  className: "column is-4 is-offset-4"
})`
  && {
    padding: 1px;
  }
`;

const Heading = styled.h1.attrs({
  className: "title is-3"
})`
  && {
    text-align: center;
    margin: 20px 0 20px 0;
  }
`;

const ErrorMessage = styled.p.attrs({
  className: "error tile"
})`
  && {
    text-align: left;
    margin: 3px 0 3px 0;
    padding: 2px 0 2px 0;
    background-color: hsl(348, 100%, 61%);
    color: white;
  }
`;

class LoginForm extends React.Component<ILoginFormState> {
  static contextTypes = {
    router: PropTypes.object
  };

  public state = {
    errors: {
      password: "",
      responseError: "",
      username: ""
    },
    loading: false,
    userInput: {
      password: "",
      username: ""
    }
  };

  onHandleSubmit = async (e: React.FormEvent<EventTarget>) => {
    e.preventDefault();

    // Validates the user input fields
    const errors = validateLoginField(this.state.userInput);

    // Update the errors propetry of the state
    this.setState({ errors });

    // Check if there is an error, if there is abort signing in.
    if (errors.password || errors.username) {
      return;
    }

    try {
      this.setState({
        loading: true
      });

      const login = await api.auth.login({
        ...this.state.userInput
      });

      if (login.data.success) {
        // TO BE REMOVED
        // Confirming user is authenticated
        const user = await api.user.queryUser();

        console.log(user);

        this.setState({
          errors: {
            password: "",
            responseError: "",
            username: ""
          },
          loading: false,
          userInput: {
            password: "",
            username: ""
          }
        });

        // Redirect the user to the home page granted access to his own content
        this.context.router.history.push("/");
      }
    } catch (error) {
      this.setState({
        errors: {
          ...this.state.errors,
          responseError: error.message
        },
        loading: false
      });
    }
  };

  onHandleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const name = e.target.name;
    const value = e.target.value;
    // Updating the userInput property of the state when input field value updates
    this.setState({
      userInput: {
        ...this.state.userInput,
        [name]: value
      }
    });
  };

  render() {
    const { errors, loading } = this.state;
    const { password, username } = this.state.userInput;
    return (
      <LoginFormParent onSubmit={e => this.onHandleSubmit(e)}>
        <Heading>Login Here</Heading>
        {errors.responseError ? (
          <ErrorMessage>{errors.responseError}</ErrorMessage>
        ) : null}
        <LoginInputField
          error={errors.username}
          loading={loading}
          name={"username"}
          label={"Username"}
          onChange={this.onHandleInputChange}
          type={"text"}
          value={username}
        />

        <LoginInputField
          error={errors.password}
          loading={loading}
          name={"password"}
          label={"Password"}
          onChange={this.onHandleInputChange}
          type={"password"}
          value={password}
        />

        <button
          disabled={loading}
          className="signIn button is-success is-fullwidth is-rounded"
        >
          {loading ? "Signing in..." : "Sign in"}
        </button>
      </LoginFormParent>
    );
  }
}

export default LoginForm;
