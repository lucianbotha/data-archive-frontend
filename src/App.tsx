import * as React from "react";
import { Mutation, Query } from "react-apollo";
import { Redirect, Route, Switch } from "react-router-dom";
import DataRequestsForm from "./components/dataRequest/DataRequestsForm";
import LoginForm, { ILoginFormCache } from "./components/LoginForm";
import NavigationBar from "./components/NavigationBar";
import RegistrationForm, {
  IRegistrationFormCache
} from "./components/RegistrationForm";
import SearchForm, { ISearchFormCache } from "./components/SearchForm";
import { LOGOUT_USER_MUTATION } from "./graphql/Mutations";
import { USER_QUERY } from "./graphql/Query";

interface IUser {
  name: string;
  username: string;
  isAdmin: () => boolean;
}

interface IProtectedRouteProps {
  component: any;
  user: IUser | null | undefined;
  [propName: string]: any;
}

interface ICache {
  loginForm: ILoginFormCache;
  registrationForm: IRegistrationFormCache;
  searchForm: ISearchFormCache;
}

/**
 * A route which redirects to the login page if the user is not logged in.
 */
function ProtectedRoute({
  component: Component,
  user,
  ...rest
}: IProtectedRouteProps) {
  return (
    <Route
      {...rest}
      render={props =>
        user ? (
          <Component {...props} />
        ) : (
          <Redirect
            to={{
              pathname: "/login",
              state: { from: props.location }
            }}
          />
        )
      }
    />
  );
}

/**
 * The data archive.
 */
class App extends React.Component<any, any> {
  private cache: ICache = {
    loginForm: {},
    registrationForm: {},
    searchForm: {}
  };

  logout = async (userLogout: any) => {
    await userLogout();
  };

  public render() {
    return (
      <Query query={USER_QUERY}>
        {({ data, loading }) => {
          if (loading) {
            return <p>Loading...</p>;
          }

          const currentUser =
            data && data.user
              ? {
                  isAdmin: () => false,
                  name: data.user.givenName,
                  username: data.user.familyName
                }
              : null;

          return (
            <>
              <Mutation
                mutation={LOGOUT_USER_MUTATION}
                refetchQueries={[{ query: USER_QUERY }]}
              >
                {userLogout => {
                  return (
                    <NavigationBar
                      user={currentUser}
                      logout={() => this.logout(userLogout)}
                    />
                  );
                }}
              </Mutation>

              <Switch>
                {/* search page */}
                <Route
                  exact={true}
                  path="/"
                  render={() => <SearchForm cache={this.cache.searchForm} />}
                />

                {/* registration page */}
                <Route
                  exact={true}
                  path="/register"
                  component={() => (
                    <RegistrationForm cache={this.cache.registrationForm} />
                  )}
                />

                {/* login page */}
                <Route
                  exact={true}
                  path="/login"
                  component={() => <LoginForm cache={this.cache.loginForm} />}
                />

                {/* account details page */}
                <ProtectedRoute
                  user={currentUser}
                  exact={true}
                  path="/account"
                  component={() => (
                    <h1 className="title">
                      {currentUser
                        ? `${currentUser.name} ${currentUser.username} Account`
                        : "No User"}
                    </h1>
                  )}
                />

                {/* data requests page */}
                <ProtectedRoute
                  user={currentUser}
                  exact={true}
                  path="/data-requests"
                  component={() => <DataRequestsForm />}
                />

                {/* cart page */}
                <Route
                  exact={true}
                  path="/cart"
                  component={() => <h1 className="title">Cart page</h1>}
                />

                {/* admin page */}
                <ProtectedRoute
                  user={currentUser}
                  exact={true}
                  path="/admin"
                  component={() => <h1 className="title">Admin page</h1>}
                />

                {/* page not found */}
                <Route
                  component={() => <h1 className="title">Page not found</h1>}
                />
              </Switch>
            </>
          );
        }}
      </Query>
    );
  }
}

export default App;
