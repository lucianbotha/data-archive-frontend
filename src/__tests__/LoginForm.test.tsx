import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import * as React from "react";
import { MockedProvider } from "react-apollo/test-utils";
import { MemoryRouter } from "react-router";
import wait from "waait";
import App from "../App";
import LoginForm from "../components/LoginForm";
import { GET_USER_MUTATION } from "../graphql/Mutations";
import { USER_QUERY } from "../graphql/Query";
import click from "../util/click";

// Helper function for simulating input field value change.
function inputTyping(wrapper: any, name: string, value: string) {
  wrapper.find(`input[name="${name}"]`).simulate("change", {
    target: { name, value }
  });
}

// Updated userInput state
const updatedState = {
  errors: {
    password: "",
    username: ""
  },
  userInput: {
    password: "securepassword",
    username: "sj"
  }
};

// login mock mutation
const mocks = [
  {
    request: {
      query: GET_USER_MUTATION,
      variables: {
        ...updatedState.userInput
      }
    },
    result: {
      data: {
        login: true
      }
    }
  },

  {
    request: {
      query: USER_QUERY
    },
    result: {
      data: {
        user: {
          __typename: "User",
          familyName: "test",
          givenName: "test",
          roles: []
        }
      }
    }
  }
];

describe("LoginForm Component", () => {
  it("renders the LoginForm component having unpopulated props with no errors", async () => {
    // LoginForm component wrapper.
    const wrapper = mount(
      <MockedProvider>
        <LoginForm />
      </MockedProvider>
    );
    // The actual form.
    const form = wrapper.find('form[data-test="form"]');
    // Expect the snapshot to match the LoginForm component.
    expect(toJson(form)).toMatchSnapshot();
  });

  it("displays no errors if submitted inputs are all valid", async () => {
    // LoginForm component wrapper.
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <MemoryRouter initialEntries={["/login"]}>
          <LoginForm />
        </MemoryRouter>
      </MockedProvider>
    );

    // LoginForm component instance
    const instance = wrapper.find("LoginForm").instance() as any;

    // Simulate state change when the username input field value changes.
    inputTyping(wrapper, "username", "sj");

    // Simulate state change when the password input field value changes.
    inputTyping(wrapper, "password", "securepassword");

    // Expect the properties username and password of the state to have been updated with the correct value.
    expect(instance.state.userInput).toMatchObject({
      password: "securepassword",
      username: "sj"
    });

    const signInButton = wrapper.find('[data-test="signIn"]');
    // Expect the button to not be clicked
    expect(signInButton.text()).toContain("Sign in");

    // Simulate the submiting of the form.
    signInButton.simulate("submit");

    // Expect the button to have beeen clicked
    expect(signInButton.text()).toContain("Signing in");

    // Expect no error message.
    expect(wrapper.find("p").length).toBe(0);
  });

  it("displays error message if submitted invalid username", async () => {
    // LoginForm component wrapper.
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <LoginForm />
      </MockedProvider>
    );

    // LoginForm component instance
    const instance = wrapper.find("LoginForm").instance() as any;

    // Set the state with valid information
    instance.setState({ userInput: updatedState.userInput });

    // Spy on the setState function.
    const setState = jest.spyOn(instance, "setState");

    // Simulate state change when the username input field value changes.
    inputTyping(wrapper, "username", "sJ");

    // Expect setState to have been called
    expect(setState.mock.calls.length).toBe(1);

    const signInButton = wrapper.find('[data-test="signIn"]');
    // Simulate the submiting of the form.
    signInButton.simulate("submit");

    // Expect the button to have beeen clicked
    expect(signInButton.text()).toContain("Sign in");

    // Expect an error message.
    expect(wrapper.find("p").length).toBe(1);

    // Expect meaningful error message
    expect(wrapper.find("p").text()).toContain("Username");
    expect(wrapper.find("p").text()).toContain("lowercase");
  });

  it("displays error message if submitted invalid password", () => {
    // LoginForm component wrapper.
    const wrapper = mount(
      <MockedProvider>
        <LoginForm />
      </MockedProvider>
    );

    // LoginForm component instance
    const instance = wrapper.find("LoginForm").instance() as any;

    // Set the state with valid information
    instance.setState({ userInput: updatedState.userInput });

    // Spy on the setState function.
    const setState = jest.spyOn(instance, "setState");

    // Simulate state change when the username input field value changes.
    inputTyping(wrapper, "password", "short");

    // Expect setState to have been called
    expect(setState.mock.calls.length).toBe(1);

    const signInButton = wrapper.find('[data-test="signIn"]');
    // Simulate the submiting of the form
    signInButton.simulate("submit");

    // Expect the button to have beeen clicked
    expect(signInButton.text()).toContain("Sign in");

    // Expect an error message.
    expect(wrapper.find("p").length).toBe(1);

    // Expect meaningful error message
    expect(
      wrapper
        .find("p")
        .first()
        .text()
    ).toContain("Password");
    expect(
      wrapper
        .find("p")
        .first()
        .text()
    ).toContain("7 characters");
  });

  it("should cache values and errors", async () => {
    const wrapper = mount(
      <MockedProvider>
        <MemoryRouter>
          <App />
        </MemoryRouter>
      </MockedProvider>
    );

    await wait(0);
    wrapper.update();

    // Navigate to the login form
    const loginFormLink = wrapper.find('a[href="/login"]').first();
    click(loginFormLink);

    await wait(0);
    wrapper.update();

    // Fill in an invalid password
    inputTyping(wrapper, "password", "short");

    // Submit the form
    const submitButton = wrapper.find('button[data-test="signIn"]');
    submitButton.simulate("submit");

    await wait(0);
    wrapper.update();

    // The value has been stored in the state, and there is an error
    const loginFormState: any = wrapper.find("LoginForm").state();
    const passwordValue = loginFormState.userInput.password;
    expect(passwordValue).toEqual("short");
    const passwordErrorMessage = loginFormState.errors.password;
    expect(passwordErrorMessage.length).toBeGreaterThan(0);

    // Navigate away from the login form
    const cartLink = wrapper.find('a[href="/cart"]').first();
    click(cartLink);

    await wait(0);
    wrapper.update();

    // No login form any longer
    expect(wrapper.find("LoginForm").length).toBe(0);

    // Navigate back to the search form
    click(loginFormLink);

    await wait(0);
    wrapper.update();

    // The values and errors have been re-inserted.
    const newLoginFormState: any = wrapper.find("LoginForm").state();
    expect(newLoginFormState.userInput.password).toEqual(passwordValue);
    expect(newLoginFormState.errors.password).toEqual(passwordErrorMessage);
  });
});
