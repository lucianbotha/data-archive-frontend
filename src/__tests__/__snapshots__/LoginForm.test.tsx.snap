// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm Component renders the LoginForm component having unpopulated props with no errors 1`] = `
.c0.c0 {
  padding: 1px;
}

.c1.c1 {
  text-align: center;
  margin: 20px 0 20px 0;
}

<LoginForm>
  <styled.form
    onSubmit={[Function]}
  >
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [
            Object {
              "className": "column is-4 is-offset-4",
            },
          ],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bwzfXH",
            "isStatic": true,
            "lastClassName": "c0",
            "rules": Array [
              "
  && {
    padding: 1px;
  }
",
            ],
          },
          "displayName": "styled.form",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bwzfXH",
          "target": "form",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      onSubmit={[Function]}
    >
      <form
        className="column is-4 is-offset-4 c0"
        onSubmit={[Function]}
      >
        <styled.h1>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [
                  Object {
                    "className": "title is-3",
                  },
                ],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-htpNat",
                  "isStatic": true,
                  "lastClassName": "c1",
                  "rules": Array [
                    "
  && {
    text-align: center;
    margin: 20px 0 20px 0;
  }
",
                  ],
                },
                "displayName": "styled.h1",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-htpNat",
                "target": "h1",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <h1
              className="title is-3 c1"
            >
              Login to the Data Archive
            </h1>
          </StyledComponent>
        </styled.h1>
        <fieldset
          aria-disabled={false}
          disabled={false}
        >
          <div
            className="field"
          >
            <label
              className="label"
            >
              Username
              <div
                className="control is-child"
              >
                <InputField
                  error=""
                  name="username"
                  onChange={[Function]}
                  type="text"
                  value=""
                >
                  <div
                    className="control"
                  >
                    <input
                      className="input undefined "
                      name="username"
                      onChange={[Function]}
                      type="text"
                      value=""
                    />
                  </div>
                </InputField>
              </div>
            </label>
          </div>
          <div
            className="field"
          >
            <label
              className="label"
            >
              Password
              <div
                className="control is-child"
              >
                <InputField
                  data-test="password-input"
                  error=""
                  name="password"
                  onChange={[Function]}
                  type="password"
                  value=""
                >
                  <div
                    className="control"
                  >
                    <input
                      className="input undefined "
                      data-test="password-input"
                      name="password"
                      onChange={[Function]}
                      type="password"
                      value=""
                    />
                  </div>
                </InputField>
              </div>
            </label>
          </div>
          <button
            className="button is-success is-fullwidth is-rounded"
            data-test="signIn"
          >
            Sign in
          </button>
        </fieldset>
      </form>
    </StyledComponent>
  </styled.form>
</LoginForm>
`;
